python_modules/dagster/dagster_tests/core_tests/resource_tests/test_struct_config_resources.py

```diff
@@ -1781,3 +1786,89 @@ def hello_world_job() -> None:
 
     assert defs.get_job_def("hello_world_job").execute_in_process().success
     assert outputs == ["foo"]
+
+
+def test_direct_op_invocation() -> None:
+    class MyResource(ConfigurableResource):
+        a_str: str
+
+    @op
+    def my_op(context, my_resource: MyResource) -> str:
+        assert my_resource.a_str == "foo"
+        return my_resource.a_str
+
+    # Just providing context is ok, we'll use the resource from the context
+    assert my_op(build_op_context(resources={"my_resource": MyResource(a_str="foo")})) == "foo"
+
+    # Providing both context and resource is not ok, because we don't know which one to use
+    with pytest.raises(
+        DagsterInvalidInvocationError,
+        match="Cannot provide resources in both context and kwargs",
+    ):
+        assert (
+            my_op(
+                context=build_op_context(resources={"my_resource": MyResource(a_str="foo")}),
+                my_resource=MyResource(a_str="foo"),
+            )
+            == "foo"
+        )
+
+    # Providing resource only as kwarg is ok, we'll use that (we still need a context though)
+    assert my_op(context=build_op_context(), my_resource=MyResource(a_str="foo")) == "foo"
+
+    @op
+    def my_op_no_context(my_resource: MyResource) -> str:
+        assert my_resource.a_str == "foo"
+        return my_resource.a_str
+
+    # Providing context is ok, we just discard it and use the resource from the context
+    assert (
+        my_op_no_context(build_op_context(resources={"my_resource": MyResource(a_str="foo")}))
+        == "foo"
+    )
+
+    # Providing resource only as kwarg is ok, we'll use that
+    assert my_op_no_context(my_resource=MyResource(a_str="foo")) == "foo"
```

Comment:
What's the defined behavior when there are multiple resource arguments? One kinda makes sense to just allow you to replace it with the context, but with multiple args it starts to become confusing.

Reply:
Added a new test which showcases the multiple resource arguments case. Also added a test for upstream op/asset inputs since that's pretty important. 

Right now we only support passing resources (1) through context or (2) through kwargs. It's too tricky to figure out what a user's intent is passing resource as args - there's no ambiguity if it's in a resource dict in context or has a key specified in kwargs at least.

Hopefully the new test cases are illustrative, I think it's a reasonable set of cases to support